### PR TITLE
First DirectoryServcies tests which runs against real server

### PR DESCRIPTION
**Pull request by tarekgh**
This is the first test collection which intended to run against real servers supports LDAP. 

Read the comment in the file LDAP.Configuration.xml to know how can enable such tests. by default, these tests will be skipped till providing the real server info in the file LDAP.Configuration.xml.

These are just initial tests which we should add more as we go. It is recommended when adding any more tests to test it against AD servers (i.e. Windows Server) and against OpenDJ too to avoid the behavior differences.

One more note, I have written every test case as self-contained so every test case will exit in a clean state of the server.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @AlexGhiondea @karelz @danmosemsft

**Comment by tarekgh** [comment]
Looks msbuild complaining 

```
error MSB4066: The attribute "CopyToOutputDirectory" in element <None> is unrecognized
```

I'll try to fix this one.

**Comment by karelz** [comment]
Very cool.
If we can create Docker image with the server, that would be awesome for even easier automation.

cc @hughbe

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            DirectoryEntry ouCoreDevs = de.Children.Add($"ou={ou}","Class");
+            ouCoreDevs.Properties["objectClass"].Value = "organizationalUnit";    // has to be organizationalUnit
+            ouCoreDevs.Properties["description"].Value = description;
+            ouCoreDevs.Properties["ou"].Value = ou;
+            ouCoreDevs.CommitChanges();
+            return ouCoreDevs;
+        }
+
+        private DirectoryEntry CreateCN(DirectoryEntry ouEntry, string cn, string description, string phone)
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: CN isn't really a "type" it stands for common name.  I would change this name to match the object type that is actually getting created.

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            using (DirectorySearcher ds = new DirectorySearcher(de))
+            {
+                ds.Filter = $"(&(objectClass=organizationalUnit)(ou={ouName}))";
+                ds.PropertiesToLoad.Add("ou");
+                SearchResult sr = ds.FindOne();
+                Assert.Equal(ouName, sr.Properties["ou"][0]);
+            }
+        }
+
+        private void SearchCNByName(DirectoryEntry de, string cnName)
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Same feedback on naming

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        {
+			using (DirectoryEntry de = new DirectoryEntry(LdapConfiguration.Configuration.LdapPath,
+														    LdapConfiguration.Configuration.UserName,
+														    LdapConfiguration.Configuration.Password,
+														    LdapConfiguration.Configuration.AuthenticationTypes))
+            {
+                DirectoryEntry rootOU = CreateOU(de, "CoreFxRootOU", "CoreFx Root OU");
+                try
+                {
+                    DirectoryEntry child1OU = CreateOU(rootOU, "CoreFxChild1OU", "CoreFx Child OU 1");
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: These need using statements

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                    DirectoryEntry child2OU = CreateOU(rootOU, "CoreFxChild2OU", "CoreFx Child OU 2");
+
+                    CreateCN(child1OU, "user.ou1.1", "User 1 is in CoreFx ou 1", "1 111 111 11111");
+                    CreateCN(child1OU, "user.ou1.2", "User 2 is in CoreFx ou 1", "1 222 222 22222");
+
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: I suppose we'll add some test logging in the future so it's clear which of these sub steps failed when the test fails?

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+				DeleteDirectoryEntry(de, child);
+			}
+			
+            parent.Children.Remove(de);
+			parent.CommitChanges();
+        }
+
+        private void SearchOUByName(DirectoryEntry de, string ouName)
+        {
+            using (DirectorySearcher ds = new DirectorySearcher(de))
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Set ds.ClientTimeout to some reasonable value.  I think the default is infinite wait.

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        [ConditionalFact(nameof(IsLdapConfigurationExist))]
+        public void TestOU() // adding and removing organization unit
+        {
+			using (DirectoryEntry de = new DirectoryEntry(LdapConfiguration.Configuration.LdapPath,
+														    LdapConfiguration.Configuration.UserName,
+														    LdapConfiguration.Configuration.Password,
+														    LdapConfiguration.Configuration.AuthenticationTypes))
+            {
+                string ouName = "NetCoreDevs";
+                CreateOU(de, ouName, ".Net Core Developers Unit");
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: You should try and remove any pre-exisiting OUs with the target name first to take care of tests which leave the environment dirty

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        private void DeleteDirectoryEntry(DirectoryEntry parent, DirectoryEntry de)
+        {
+			foreach (DirectoryEntry child in de.Children)
+            {
+				DeleteDirectoryEntry(de, child);
+			}
+			
+            parent.Children.Remove(de);
```
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: bad formatting

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 2 above)
Comment (tquerec) [inline src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: For all of these test* methods, should we add a finally block and then assert if any exception is thrown?  That way you know which action failed.

File: src/Common/tests/System/DirectoryServices/LdapConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        internal static LdapConfiguration Configuration =>  s_ldapConfiguration;
+
+        internal string ServerName { get; set; }
+        internal string UserName { get; set; }
+        internal string Password { get; set; }
+        internal string Port { get; set; }
+        internal string Domain { get; set; }
+        internal AuthenticationTypes AuthenticationTypes { get; set; }
+        internal string LdapPath => String.IsNullOrEmpty(Port) ? $"LDAP://{ServerName}/{Domain}" : $"LDAP://{ServerName}:{Port}/{Domain}";
```
Comment (tquerec) [inline src/Common/tests/System/DirectoryServices/LdapConfiguration.cs]: This format is very specific to the System.DirectoryServices* assembly but doesn't apply to System.DirectoryServices.Protocols  We may want to reuse this config class for all the S.DS.* assemblies to maybe this assembly specific logic should be moved to the assembly specific tests.

Comment (tarekgh) [reply src/Common/tests/System/DirectoryServices/LdapConfiguration.cs]: When adding tests to System.DirectoryServices.Protocols, we can look at what configuration is used there and then can update the configuration class to support it. or to use another specific configuration class.

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: This is not needed. XUnit is catching the thrown exception and fail the test with the exception logs. we should be fine here.

Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 5 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that. I'll use 2 minutes timeout which I think can be reasonable.

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 7 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 1 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: XUnit will catch any thrown exception and have it in the logs.

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 3 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that

File: src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs (hunk 6 above)
Comment (tarekgh) [reply src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs]: Thanks. I'll fix that

**Comment by tarekgh** [comment]
Thanks @tquerec for your review

**Pull request merged**